HIP/hip-435.md

```diff
@@ -0,0 +1,131 @@
+
+## Rationale
+
+There are use-cases requiring a detailed and verbose output for a HAPI transaction execution. Users of Hedera should have the option to query a mirror node for this information. Meanwhile, bloating the TransactionRecords with information that would be unnecessary in most cases is not an ideal option. It is a much better design decision to create sidecar records that can be downloaded and processed only by interested parties.
+
+**Design Goal #1**
+Mirror nodes should not be required to download and process verbose transaction information. The sidecar transaction records will be externalised in separate record files.
+
+**Design Goal #2**
+One transaction can have 0, 1 or more than 1 sidecar records related to it.
+
+**Design Goal #3**
+It should be possible and relatively easy to add new types of sidecar transaction records.
+
+## User stories
+
+As a Hedera user, I want the to have the ability to access verbose information about transactions.
+
+As an operator of a mirror node, I want the ability to download and process only the most important record-related information.
+
+As an operator of a mirror node, I want the ability to download all the information about records execution that a Hedera node can provide.
+
+## Specification
+
+Introduce a sidecar transaction record protobuf message and a sidecar record file protobuf message to enable tracking of additional information about records execution.
+
+### TransactionSideCarRecord
+
+`TransactionSideCarRecord` is a protobuf which will be used to create sidecar records complementing `TransactionRecord` and storing additional information about a transaction execution.
+The `TransactionSideCarRecord`s created by Hedera nodes will be externalised in a new file, part of the new RecordStream structure described bellow.
+
+```
+message TransactionSideCarRecord {
+    // Consensus timestamp will be the same as the consensus timestamp of the 
+    // transaction the side car is related to. This offers a convenient 
+    // way to match record to sidecar. 
+    Timestamp consensusTimestamp = 1;
+    oneof {
+      repeated ContractStateChange = 2; // Example sidecar type
+      repeated ContractAction = 3; // Example sidecar type
+      // Other future categories
+    }
+}
+```
+
+### Record Stream File Changes
+
+A new version (V6) of the record stream file format will be created using the following protobufs:
+
+```
+message RecordFileObject {
+  TransactionRecord record = 1;
+  Transaction transaction = 2;
+}
+
+message RecordStreamFile {
+  bytes startObjectRunningHash = 1;
+  repeated RecordFileObject fileObjects = 2;
+  bytes endObjectRunningHash = 3;
+  uint64 blockNumber = 4;
+  repeated bytes SideCarFileHash = 5; //allows multiple sidecar files to be linked to this RecordStreamFile
+}
+
+message SideCarFile {
```

Comment:
Do we want only one record per sidecar file?  Or is the sidecar file a stream of side car records?

Reply:
The SideCarFile will contain all sidecar records of a given type (e.g. ContractAction, StateChanges) that were produced during the same ~2s period that the record file was produced. In other words all the sidecar records relating to transaction records in the record file for that period. This also means that multiple SideCarFiles will be produced per RecordStreamFile - 1 for ContractAction sidecars, 1 for StateChanges, 1 for each new type of sidecar records we introduce in the future.